diff --git a/CMakeLists.txt b/CMakeLists.txt
new file mode 100644
index 0000000..3f1693b
--- /dev/null
+++ b/CMakeLists.txt
@@ -0,0 +1,28 @@
+project(baumwelch)
+cmake_minimum_required(VERSION 3.18)
+
+include(GNUInstallDirs)
+include(GenerateExportHeader)
+set(CMAKE_MACOSX_RPATH 1)
+set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_CXX_EXTENSIONS OFF)
+
+set(CMAKE_POSITION_INDEPENDENT_CODE ON)
+if (MSVC)
+    find_package(dlfcn-win32 REQUIRED)
+    set(CMAKE_DL_LIBS dlfcn-win32::dl)
+    set(CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS ON)
+    add_definitions(-DWIN32_LEAN_AND_MEAN)
+    add_definitions(-DNOMINMAX)
+    add_definitions(-D_SILENCE_ALL_CXX17_DEPRECATION_WARNINGS)
+    add_definitions(-D_USE_MATH_DEFINES)
+    add_compile_options(/permissive- /FS /wd4819 /EHsc /bigobj)
+
+    # some warnings related with fst
+    add_compile_options(/wd4018 /wd4244 /wd4267 /wd4291 /wd4305)
+endif (MSVC)
+option(BUILD_SHARED_LIBS "Build shared libraries" ON)
+
+set(SOVERSION "36")
+
+add_subdirectory(src)
diff --git a/src/bin/CMakeLists.txt b/src/bin/CMakeLists.txt
new file mode 100644
index 0000000..7d8e691
--- /dev/null
+++ b/src/bin/CMakeLists.txt
@@ -0,0 +1,16 @@
+function (add_executable2 _name)
+    add_executable(${ARGV})
+    if (TARGET ${_name})
+        target_link_libraries(${_name} baumwelchscript fstfarscript fstfar fstscript fst ${CMAKE_DL_LIBS})
+    endif()
+
+    install(TARGETS ${_name} RUNTIME DESTINATION bin)
+endfunction()
+
+include_directories(../include)
+
+add_executable2(baumwelchdecode baumwelchdecode-main.cc baumwelchdecode.cc)
+
+add_executable2(baumwelchrandomize baumwelchrandomize-main.cc baumwelchrandomize.cc)
+
+add_executable2(baumwelchtrain  baumwelchtrain-main.cc baumwelchtrain.cc)
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
new file mode 100644
index 0000000..5b8cff2
--- /dev/null
+++ b/src/CMakeLists.txt
@@ -0,0 +1,7 @@
+include_directories(./include/)
+install(DIRECTORY include/ DESTINATION include/
+        FILES_MATCHING PATTERN "*.h")
+
+add_subdirectory(script)
+
+add_subdirectory(bin)
\ No newline at end of file
diff --git a/src/include/baumwelch/baumwelchscript_Export.h b/src/include/baumwelch/baumwelchscript_Export.h
new file mode 100644
index 0000000..8ea1f55
--- /dev/null
+++ b/src/include/baumwelch/baumwelchscript_Export.h
@@ -0,0 +1,41 @@
+#ifndef baumwelchscript_EXPORT_H
+#define baumwelchscript_EXPORT_H
+
+#ifdef baumwelchscript_BUILT_AS_STATIC
+#  define baumwelchscript_EXPORT
+#  define BAUMWELCHSCRIPT_NO_EXPORT
+#else
+#  ifndef baumwelchscript_EXPORT
+#    ifdef baumwelchscript_EXPORTS
+        /* We are building this library */
+#      define baumwelchscript_EXPORT __declspec(dllexport)
+#    else
+        /* We are using this library */
+#      define baumwelchscript_EXPORT __declspec(dllimport)
+#    endif
+#  endif
+
+#  ifndef BAUMWELCHSCRIPT_NO_EXPORT
+#    define BAUMWELCHSCRIPT_NO_EXPORT
+#  endif
+#endif
+
+#ifndef BAUMWELCHSCRIPT_DEPRECATED
+#  define BAUMWELCHSCRIPT_DEPRECATED __declspec(deprecated)
+#endif
+
+#ifndef BAUMWELCHSCRIPT_DEPRECATED_EXPORT
+#  define BAUMWELCHSCRIPT_DEPRECATED_EXPORT baumwelchscript_EXPORT BAUMWELCHSCRIPT_DEPRECATED
+#endif
+
+#ifndef BAUMWELCHSCRIPT_DEPRECATED_NO_EXPORT
+#  define BAUMWELCHSCRIPT_DEPRECATED_NO_EXPORT BAUMWELCHSCRIPT_NO_EXPORT BAUMWELCHSCRIPT_DEPRECATED
+#endif
+
+#if 0 /* DEFINE_NO_DEPRECATED */
+#  ifndef BAUMWELCHSCRIPT_NO_DEPRECATED
+#    define BAUMWELCHSCRIPT_NO_DEPRECATED
+#  endif
+#endif
+
+#endif /* baumwelchscript_EXPORT_H */
\ No newline at end of file
diff --git a/src/include/baumwelch/decodescript.h b/src/include/baumwelch/decodescript.h
index c6b56ff..7592880 100644
--- a/src/include/baumwelch/decodescript.h
+++ b/src/include/baumwelch/decodescript.h
@@ -22,6 +22,7 @@
 #include <fst/fst.h>
 #include <fst/script/fst-class.h>
 #include <baumwelch/decode.h>
+#include <baumwelch/baumwelchscript_Export.h>
 
 namespace fst {
 namespace script {
@@ -41,7 +42,7 @@ void Decode(BaumWelchDecodeArgs *args) {
   Decode(input, output, model, hypotext, mapper);
 }
 
-void Decode(FarReaderClass& input, FarReaderClass& output,
+void baumwelchscript_EXPORT Decode(FarReaderClass& input, FarReaderClass& output,
             const FstClass& model, FarWriterClass& hypotext,
             EncodeMapperClass* encoder = nullptr);
 
diff --git a/src/include/baumwelch/randomizescript.h b/src/include/baumwelch/randomizescript.h
index 24182e4..cd1500c 100644
--- a/src/include/baumwelch/randomizescript.h
+++ b/src/include/baumwelch/randomizescript.h
@@ -22,6 +22,7 @@
 #include <fst/mutable-fst.h>
 #include <fst/script/fst-class.h>
 #include <baumwelch/randomize.h>
+#include <baumwelch/baumwelchscript_Export.h>
 
 namespace fst {
 namespace script {
@@ -34,7 +35,7 @@ void Randomize(BaumWelchRandomizeArgs *args) {
   Randomize(model, std::get<1>(*args));
 }
 
-void Randomize(MutableFstClass *fst, uint64_t seed = std::random_device()());
+void baumwelchscript_EXPORT Randomize(MutableFstClass *fst, uint64_t seed = std::random_device()());
 
 }  // namespace script
 }  // namespace fst
diff --git a/src/include/baumwelch/trainscript.h b/src/include/baumwelch/trainscript.h
index 8d963c7..65334f1 100644
--- a/src/include/baumwelch/trainscript.h
+++ b/src/include/baumwelch/trainscript.h
@@ -22,6 +22,7 @@
 #include <fst/mutable-fst.h>
 #include <fst/script/fst-class.h>
 #include <baumwelch/train.h>
+#include <baumwelch/baumwelchscript_Export.h>
 
 namespace fst {
 namespace script {
@@ -38,7 +39,7 @@ void Train(BaumWelchTrainArgs *args) {
   Train(input, output, model, std::get<3>(*args), std::get<4>(*args));
 }
 
-void Train(FarReaderClass &input, FarReaderClass &output,
+void baumwelchscript_EXPORT Train(FarReaderClass &input, FarReaderClass &output,
            MutableFstClass *model, bool normalize_ilabel = true,
            const TrainOptions &opts = TrainOptions());
 
diff --git a/src/script/CMakeLists.txt b/src/script/CMakeLists.txt
new file mode 100644
index 0000000..6e03b89
--- /dev/null
+++ b/src/script/CMakeLists.txt
@@ -0,0 +1,18 @@
+
+add_library(baumwelchscript
+SHARED
+decodescript.cc
+randomizescript.cc
+trainscript.cc
+)
+target_link_libraries(baumwelchscript PUBLIC fst fstfar fstscript fstfarscript ${CMAKE_DL_LIBS})
+target_include_directories(baumwelchscript PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+                         $<INSTALL_INTERFACE:include>)
+
+set_target_properties(baumwelchscript PROPERTIES
+SOVERSION "${SOVERSION}"
+)
+install(TARGETS baumwelchscript
+LIBRARY DESTINATION lib
+ARCHIVE DESTINATION lib
+RUNTIME DESTINATION bin)
